python/mxnet/rnn/rnn_cell.py

```diff
@@ -0,0 +1,172 @@
+# coding: utf-8
+"""Definition of various recurrent neural network cells."""
+
+import copy
+
+from .. import symbol
+from .. import ndarray
+from ..base import numeric_types, string_types
+
+class RNNParams(object):
+    def __init__(self):
+        self._params = {}
+
+    def get(self, name, **kwargs):
+        if name not in self._params:
+            self._params[name] = symbol.Variable(name, **kwargs)
+        return self._params[name]
+
+
+class BaseRNNCell(object):
+    """Abstract base class for RNN cells"""
+    def __call__(self, inputs, states, params, prefix=''):
+        """construct symbol"""
+        raise NotImplementedError()
+
+    @property
+    def state_shape(self):
+        """shape(s) of states"""
+        raise NotImplementedError()
+
+    @property
+    def output_shape(self):
+        """shape(s) of output"""
+        raise NotImplementedError()
+
+    def begin_state(self, prefix='', init_sym=symbol.zeros, **kwargs):
+        """initial state"""
+        state_shape = self.state_shape
+        def recursive(shape, c):
```

Comment:
I don't understand what this is doing.  Can you add a comment -- It looks like the state_shape property is overloaded to support multiple different types.  We should document what the options are and why.